src/plugins/doi.ts

```diff
@@ -0,0 +1,70 @@
+/*!
+ * © 2024 Atypon Systems LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { schema } from '@manuscripts/transform'
+import { Node } from 'prosemirror-model'
+import { Plugin } from 'prosemirror-state'
+import { Decoration, DecorationSet } from 'prosemirror-view'
+
+export default () => {
+  return new Plugin({
+    state: {
+      init(_, { doc }) {
+        const decorations = createDecoration(doc)
+        return DecorationSet.create(doc, decorations)
+      },
+      apply(tr, oldDecorationSet, oldState, newState) {
+        if (tr.docChanged) {
+          const decorations = createDecoration(newState.doc)
+          return DecorationSet.create(newState.doc, decorations)
+        }
+        return oldDecorationSet
+      },
+    },
+    props: {
+      decorations(state) {
+        return this.getState(state)
+      },
+    },
+  })
+}
+
+function getPosition(doc: Node) {
+  let position = 0
+  const possibleNodesTypes = ['keywords', 'supplements', 'abstracts', 'body']
+  for (const type of possibleNodesTypes) {
+    doc.descendants((node, pos) => {
+      if (node.type === schema.nodes[type]) {
```

Comment:
while looping over nodes in the .descendants method you can record position for each "possibleNodesTypes" and then just grab the most relevant value from the result array. That way you won't have to go over the descendants more than once, which can take sometime in a larger document.

Reply:
updated... positons are saved in array and then the minimum one is selected...